// fp-ts-std/dist/cjs/IOEither.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pass = exports.traverseArray_ = exports.sequenceArray_ = exports.unsafeExpectLeft = exports.unsafeExpect = exports.unsafeUnwrapLeft = exports.unsafeUnwrap = void 0;
const IOE = require("fp-ts/IOEither");
const IO = require("fp-ts/IO");
const IO_1 = require("./IO");
const Either_1 = require("./Either");
const function_1 = require("fp-ts/function");
const Applicative_1 = require("./Applicative");
exports.unsafeUnwrap = (0, function_1.flow)(IO.map(Either_1.unsafeUnwrap), IO_1.execute);
exports.unsafeUnwrapLeft = (0, function_1.flow)(IO.map(Either_1.unsafeUnwrapLeft), IO_1.execute);
const unsafeExpect = (S) => (0, function_1.flow)(IOE.mapLeft(S.show), exports.unsafeUnwrap);
exports.unsafeExpect = unsafeExpect;
const unsafeExpectLeft = (S) => (0, function_1.flow)(IOE.map(S.show), exports.unsafeUnwrapLeft);
exports.unsafeExpectLeft = unsafeExpectLeft;
exports.sequenceArray_ = (0, function_1.flow)(IOE.sequenceArray, IOE.map(function_1.constVoid));
const traverseArray_ = f => (0, function_1.flow)(IOE.traverseArray(f), IOE.map(function_1.constVoid));
exports.traverseArray_ = traverseArray_;
exports.pass = (0, Applicative_1.pass)(IOE.ApplicativePar);
