// fp-ts-std/dist/cjs/TaskEither.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pass = exports.traverseSeqArray_ = exports.traverseArray_ = exports.sequenceSeqArray_ = exports.sequenceArray_ = exports.unsafeExpectLeft = exports.unsafeExpect = exports.unsafeUnwrapLeft = exports.unsafeUnwrap = void 0;
const TE = require("fp-ts/TaskEither");
const T = require("fp-ts/Task");
const Task_1 = require("./Task");
const Either_1 = require("./Either");
const function_1 = require("fp-ts/function");
const Applicative_1 = require("./Applicative");
exports.unsafeUnwrap = (0, function_1.flow)(T.map(Either_1.unsafeUnwrap), Task_1.execute);
exports.unsafeUnwrapLeft = (0, function_1.flow)(T.map(Either_1.unsafeUnwrapLeft), Task_1.execute);
const unsafeExpect = (S) => (0, function_1.flow)(TE.mapLeft(S.show), exports.unsafeUnwrap);
exports.unsafeExpect = unsafeExpect;
const unsafeExpectLeft = (S) => (0, function_1.flow)(TE.map(S.show), exports.unsafeUnwrapLeft);
exports.unsafeExpectLeft = unsafeExpectLeft;
exports.sequenceArray_ = (0, function_1.flow)(TE.sequenceArray, TE.map(function_1.constVoid));
exports.sequenceSeqArray_ = (0, function_1.flow)(TE.sequenceSeqArray, TE.map(function_1.constVoid));
const traverseArray_ = f => (0, function_1.flow)(TE.traverseArray(f), TE.map(function_1.constVoid));
exports.traverseArray_ = traverseArray_;
const traverseSeqArray_ = f => (0, function_1.flow)(TE.traverseSeqArray(f), TE.map(function_1.constVoid));
exports.traverseSeqArray_ = traverseSeqArray_;
exports.pass = (0, Applicative_1.pass)(TE.ApplicativePar);
