// spectacles-ts/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rename = exports.remove = exports.upsert = exports.modifyF = exports.modifyOptionW = exports.modifyW = exports.modifyOption = exports.modify = exports.setOption = exports.set = exports.get = void 0;
const get_1 = require("./values/get");
Object.defineProperty(exports, "get", { enumerable: true, get: function () { return get_1.get; } });
const set_1 = require("./values/set");
Object.defineProperty(exports, "set", { enumerable: true, get: function () { return set_1.set; } });
const setOption_1 = require("./values/setOption");
Object.defineProperty(exports, "setOption", { enumerable: true, get: function () { return setOption_1.setOption; } });
const modify_1 = require("./values/modify");
Object.defineProperty(exports, "modify", { enumerable: true, get: function () { return modify_1.modify; } });
const modifyOption_1 = require("./values/modifyOption");
Object.defineProperty(exports, "modifyOption", { enumerable: true, get: function () { return modifyOption_1.modifyOption; } });
const modifyW_1 = require("./values/modifyW");
Object.defineProperty(exports, "modifyW", { enumerable: true, get: function () { return modifyW_1.modifyW; } });
const modifyOptionW_1 = require("./values/modifyOptionW");
Object.defineProperty(exports, "modifyOptionW", { enumerable: true, get: function () { return modifyOptionW_1.modifyOptionW; } });
const modifyF_1 = require("./values/modifyF");
Object.defineProperty(exports, "modifyF", { enumerable: true, get: function () { return modifyF_1.modifyF; } });
const upsert_1 = require("./values/upsert");
Object.defineProperty(exports, "upsert", { enumerable: true, get: function () { return upsert_1.upsert; } });
const remove_1 = require("./values/remove");
Object.defineProperty(exports, "remove", { enumerable: true, get: function () { return remove_1.remove; } });
const rename_1 = require("./values/rename");
Object.defineProperty(exports, "rename", { enumerable: true, get: function () { return rename_1.rename; } });
